### PR TITLE
Phil

### PR DESCRIPTION
**Pull request by philsof**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,5 @@
 
 # Ignore Byebug command history file.
 .byebug_history
+
+/*.db
```
Comment (philsof) [inline .gitignore:19]: Check out [this repository](https://github.com/github/gitignore) for best-practice .gitignore files for many types of projects (Rails, etc.) and also [many types of systems](https://github.com/github/gitignore/tree/master/Global) (macOS, etc.)

File: Gemfile [hunk 2]
```diff
@@ -4,27 +4,33 @@ git_source(:github) do |repo_name|
...
 # gem 'therubyracer', platforms: :ruby
 
 # Use jquery as the JavaScript library
 gem 'jquery-rails'
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
 gem 'jbuilder', '~> 2.5'
+gem 'bcrypt'
+gem 'rest-client'
+gem 'pry'
+gem 'pry-byebug'
```
Comment (philsof) [inline Gemfile:31]: Gems that are used for debugging should not be available in production. Put` gem pry` and `gem pry-byebug` in `group :development, :test` to ensure they are not available in production.

File: app/assets/javascripts/components/SearchBar.es6.jsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    $.ajax({
+      url: "/search",
+      method: "POST",
+      data: {searchResult: this.state.value}
+    })
+    .done(function(response){
+      this.props.addResult(response);
+    }.bind(this))
+    .fail(function(response){
+      // console.log(response)
```
Comment (philsof) [inline app/assets/javascripts/components/SearchBar.es6.jsx:26]: With more time, error messaging would be taken care of here.

File: app/assets/javascripts/components/SearchContainer.es6.jsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+class SearchContainer extends React.Component {
+
+  constructor(props){
+    super(props);
+    this.state = {searchResults: []}
```
Comment (philsof) [inline app/assets/javascripts/components/SearchContainer.es6.jsx:5]: Heads up: doesn't look like you ever reference `SearchContainer.state` or `SearchContainer.setState` in your code. If so, this line is not needed.

File: app/assets/javascripts/components/SearchContainer.es6.jsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  }
+
+  render(){
+    // debugger
```
Comment (philsof) [inline app/assets/javascripts/components/SearchContainer.es6.jsx:9]: Be sure to remove debugging code from final product.

File: app/assets/javascripts/components/SearchContainer.es6.jsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    return (
+      <div>
+        <h1>Movies</h1>
+          <ul className="collection">
+            {
+              this.props.searchResults.map((data, i) =>{
+                return <SearchResult data={data} key={i}/>
```
Comment (philsof) [inline app/assets/javascripts/components/SearchContainer.es6.jsx:16]: `=>` will work here, but be aware that this could also be written as a traditional `function()`. For a good explanation of the difference between `=>` and `function` check [this out](https://frontendmasters.com/courses/es6-right-parts/#v=qc7w3t7dcw).

File: app/assets/javascripts/components/SearchResult.es6.jsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
+class SearchResult extends React.Component {
+
+  constructor(props){
+    super(props);
+  }
+
+  render(){
+    return (
+      <li className="collection-item avatar">
+      <img src={this.props.data.Poster} alt="" className="circle" />
```
Comment (philsof) [inline app/assets/javascripts/components/SearchResult.es6.jsx:10]: Let's follow naming conventions here `this.props.data.poster`

File: app/assets/javascripts/components/SearchResult.es6.jsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      <span className="title"><a href={"/movies/" + this.props.data.id}>{this.props.data.Title}</a> </span>
```
Comment (philsof) [inline app/assets/javascripts/components/SearchResult.es6.jsx:11]: same here

File: app/assets/stylesheets/conversations.scss [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+
+img.card-content{
+  margin-top: 30px;
+}
+
+.new_conversation{
+  // border: 1px solid black;
+  float: center;
+margin-left: 15%;
```
Comment (philsof) [inline app/assets/stylesheets/conversations.scss:35]: Watch indentation

File: app/controllers/comments_controller.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  include ApplicationHelper
+
+  # def new
+  #   @comment = Post.new(user_id: current_user.id, movie_id: params[:user_id])
+  # end
+
+  def create
+    @comment = Comment.new(comment_params)
+    @movie = @comment.movie
+    @comments = @movie.comments
```
Comment (philsof) [inline app/controllers/comments_controller.rb:11]: Check this out: Right now, lines 10-11 are making two SQL queries (line 10 queries the database for the movie associated with the comment, and line 11 queries the database for the comments associated with the movie.) If we can get all of this info with one query, we can cut our query time in half, most likely.  Can we do this? Yes:

```ruby
@movie = @comment.movie.includes(:comments)
```
This is called eager loading. Rails docs [here](http://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations). 

Added bonus: now, in your view, if you call `@movie.comments`, there is no need to query the database again, because @movie already contains the comments associated with the movie.

File: app/controllers/comments_controller.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    else
+      @errors = ["Comment cannot be empty"]
+      render 'movies/show'
+    end
+  end
+
+  def edit
+    @comment = Comment.find_by(id: params[:id])
+    @movie =@comment.movie
+    @comments = @movie.comments
```
Comment (philsof) [inline app/controllers/comments_controller.rb:24]: Let's eager load all of this with one query (instead of three queries):
```ruby
@comment = Comment.find_by(id: params[:id]).includes(:movie).includes(:comments)
```
Now, in your view, pull everything out of the `@comment`

Doing it like this shows understanding of query optimization. That is a good thing.

(Also: this applies to every other place in your code where you are making multiple queries for associated records.)

File: app/controllers/comments_controller.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    @conversation = Conversation.new
+    render 'movies/show'
+  end
+
+  def update
+    @comment = Comment.find(params[:id])
+    @movie =@comment.movie
+    @comments = @movie.comments
+    @conversation = Conversation.new
```
Comment (philsof) [inline app/controllers/comments_controller.rb:33]: Also be aware: these variables should be instance variables (variables that start with `@`) _only_ if they are being utilized by another part of your code (e.g. the `render` call below). Remember: `redirect` cannot receive any state, since `redirect` is initiating a completely new HTTP request/response cycle. Thus, if any of these instance variables are not being utilized by the `render` call, they should be changed to local variables.

File: app/controllers/comments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+  def edit
+    @comment = Comment.find_by(id: params[:id])
+    @movie =@comment.movie
+    @comments = @movie.comments
+    @conversation = Conversation.new
+    render 'movies/show'
+  end
+
+  def update
```
Comment (philsof) [inline app/controllers/comments_controller.rb:29]: Be sure to protect any update routes

File: app/controllers/comments_controller.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
+class CommentsController < ApplicationController
+  include ApplicationHelper
+
+  # def new
+  #   @comment = Post.new(user_id: current_user.id, movie_id: params[:user_id])
+  # end
+
+  def create
+    @comment = Comment.new(comment_params)
```
Comment (philsof) [inline app/controllers/comments_controller.rb:9]: Be sure to get the new comment's user from `current_user`, not from the form (since the form data can be spoofed with a command line tool like cURL).
```ruby
@comment = Comment.new(comment_params)
@comment.user = current_user
```

File: app/controllers/comments_controller.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    @comments = @movie.comments
+    @conversation = Conversation.new
+    if @comment.update(comment_params)
+      redirect_to movie_path(@movie)
+    else
+      render 'movies/show'
+    end
+  end
+
+  def destroy
```
Comment (philsof) [inline app/controllers/comments_controller.rb:41]: Be sure to protect any destroy routes

File: app/controllers/comments_controller.rb [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    @comment = Comment.find(params[:id])
+    @comment.destroy
+    redirect_to movie_path(@comment.movie)
+  end
+
+
+  private
+  def comment_params
+    params.require(:comment).permit(:content, :movie_id, :user_id)
```
Comment (philsof) [inline app/controllers/comments_controller.rb:50]: Don't permit `user_id` here; that should come from `current_user`, since `current_user` is your only source for the _authenticated_ user.

(This note applies to every strong params method that allows user_id from a form submission.)

File: app/controllers/conversations_controller.rb [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+  def edit
+    @conversation = Conversation.find_by(id: params[:id])
+  end
+
+  def destroy
+    @conversation = Conversation.find(params[:id])
+    @movie =@conversation.comment.movie
+    @conversation.destroy
+    redirect_to movie_path(@movie)
```
Comment (philsof) [inline app/controllers/conversations_controller.rb:25]: Heads up: none of these instance variables should be instance variables, since none of them are being passed along. Let's not expose anything more than it needs to be.

File: app/controllers/movies_controller.rb [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+  def search
+    if request.xhr?
+      @movies = get_movies_by_title(params["searchResult"])
+    else
+      @movies = get_movies_by_title(params['search_form'])
+    end
+
+    if @movies
+      @movies.map! do |movie|
+        movie = Movie.find_by(imdbID: movie.imdbID)
```
Comment (philsof) [inline app/controllers/movies_controller.rb:28]: No need to reassign `movie` here, since
```ruby
Movie.find_by(imdbID: movie.imdbID)
```
 will return the exact same value as
```ruby
movie = Movie.find_by(imdbID: movie.imdbID)
```

File: app/controllers/movies_controller.rb [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+      end
+    end
+
+
+    if @movies && !request.xhr?
+      activity_logger('search', params['search_form'])
+      render "movies/index"
+    elsif request.xhr?
```
Comment (philsof) [inline app/controllers/movies_controller.rb:36]: How can we avoid calling `request.xhr?` twice in this controller action?

File: app/controllers/movies_controller.rb [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+      render :json => @movies
+    else
+      @errors = 'Movie not found' # need to implement error message partials
```
Comment (philsof) [inline app/controllers/movies_controller.rb:39]: Heads up: your errors partial view expects `@errors` to be an array.

File: app/controllers/movies_controller.rb [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+
+  end
+
+  def like
+    require_user
+    @movie = find_movie(params[:id])
+    like = @movie.likes.create(user_id: session[:user_id])
+
+    activity_logger('like', @movie) if like.valid?
+    redirect_to :back
```
Comment (philsof) [inline app/controllers/movies_controller.rb:51]: From a UX standpoint, liking a movie is a good place to implement asychronous HTTP requests on the front end, instead of hitting a redirect on the server. (Your users are not going to expect the page to reload when they click on a like button.) (Note: same with unliking a movie.)

File: app/controllers/movies_controller.rb [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+  end
+
+  def destroy_like
+    @movie = Movie.find(params[:id])
+    @like = Like.where(movie_id: @movie.id, user_id: session[:user_id])
```
Comment (philsof) [inline app/controllers/movies_controller.rb:56]: Access the authenticated user's info with `current_user`:
```ruby
@like = Like.where(movie_id: @movie.id, user: current_user)
```
(This note applies to every time the session is directly accessed in your controllers or views, except for two times: when the session is created, and when the session is destroyed.  Every other time you need to access the authenticated user, you should do so thru `current_user`.)

Comment (philsof) [reply app/controllers/movies_controller.rb:56]: Leverage your Active Record association methods:
```ruby
@like = @movie.likes.find_by(user: current_user)
```

File: app/controllers/movies_controller.rb [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+    @like.first.destroy
+    redirect_to :back
+  end
+
+  def watchlist
+    require_user
+    @movie = find_movie(params[:id])
+    watchlist = @movie.watchlists.create(user_id: session[:user_id], movie_id: params[:movie_id])
```
Comment (philsof) [inline app/controllers/movies_controller.rb:64]: 👍  leveraging the AR association `movie.watchlists.create`

File: app/controllers/movies_controller.rb [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+
+    activity_logger('watchlist', @movie) if watchlist.valid?
+    redirect_to :back
+  end
+
+  def delete_from_watchlist
+    movie = Movie.find_by(id: params[:id])
+    @entry = Watchlist.where(movie_id: movie.id, user_id: current_user.id)
```
Comment (philsof) [inline app/controllers/movies_controller.rb:72]: Make this `find_by`, which will return only the first record that matches the query.

File: app/controllers/movies_controller.rb [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+
+    @entry.first.destroy
```
Comment (philsof) [inline app/controllers/movies_controller.rb:74]: Then, this turns into `@entry.destroy`

File: app/controllers/users_controller.rb [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+      render 'new'
+    end
+
+  end
+
+  def show
+    require_user
+    @user = User.find(params[:id])
+    @favorited_movies = @user.favorited_movies
+    @watchlist_movies = @user.watchlist_movies
```
Comment (philsof) [inline app/controllers/users_controller.rb:25]: Utilize eager loading here, to obtain all of this info with one query.

File: app/helpers/application_helper.rb [hunk 27]
```diff
@@ -1,2 +1,91 @@
...
+  def get_movie_by_id( id_to_find )
+    url = "http://www.omdbapi.com"
+
+    response = RestClient.get url, {params: {i: id_to_find}}
+    parsed_response = parser(response.body)
+    @movie = Movie.find_by(imdbID: parsed_response['imdbID'])
+
+    if parsed_response['Response'] == "True"
+      @movie.update(parsed_response)
+    elsif
```
Comment (philsof) [inline app/helpers/application_helper.rb:49]:
```ruby
else
```

File: app/models/like.rb [hunk 28]
```diff
@@ -0,0 +1,6 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :movie
+
+  validates :user_id, uniqueness: {scope: :movie_id}
```
Comment (philsof) [inline app/models/like.rb:5]:
```ruby
# this will also ensure the associated user and movie are actually in the database
validates :user, uniqueness: {scope: :movie}
```
(This note applies everywhere the id of an object is validated rather than the existence of the object in the database.)

File: app/models/movie.rb [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+  validates :imdbID, uniqueness: true
+
+  after_create :update_movie_info
+
+  def related_movies
+    first_word = self.Title.split(' ')[0]
+    if first_word == 'The' || first_word == 'the'
+      first_word = self.Title.split(' ')[1]
+    end
+    related_pictures = Movie.where("movies.Title LIKE ?", "%" + first_word + "%").limit(7)
```
Comment (philsof) [inline app/models/movie.rb:18]: Postgres has cool stuff you could utilize to do this. Check out [pg_search](https://github.com/Casecommons/pg_search)

File: app/models/user.rb [hunk 30]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  has_many :comments
+  has_many :likes
+  has_many :favorited_movies, through: :likes, source: :movie
+  has_many :watchlists
+  has_many :watchlist_movies, through: :watchlists, source: :movie
+
+  has_secure_password
+  validates :name, :email, presence: true
```
Comment (philsof) [inline app/models/user.rb:9]: Definitely need some sort of uniqueness validation for your users (either on email or name). Otherwise, some users may not be able to log in!

**Review by philsof — COMMENTED** [review]
@OniDotun123 @prostko @k-shetty @TatianaAb @jordanbeth These are my comments up to the end of your models.  I will send more comments tomorrow on your views and migrations. Good stuff overall 👍

File: app/views/movies/show.html.erb [hunk 31]
```diff
@@ -0,0 +1,164 @@
+<!-- Movie title will be place here. Probably wrapped in header tags. Will be able to display the title when information is parsed from OMDB API -->
+
+
+<!-- This is the section for individual movie information. The parsed data should be able to be place into each li element once the response attributes are finalized.-->
+<section class= "center", id="movie-info" style="margin-bottom:50px;">
+  <div class="container" >
+    <ul id="movie_title">
+      <h2><%= @movie.Title %></h2>
```
Comment (philsof) [inline app/views/movies/show.html.erb:8]: `@movie.title` is Ruby naming convention (same note for all occurrences below)

File: app/views/movies/show.html.erb [hunk 32]
```diff
@@ -0,0 +1,164 @@
...
+      </ul>
+      <ul id="country">
+        <li class=""><em>Language: </em><%= @movie.Language %></li>
+        <li class=""><em>Country: </em><%= @movie.Country %></li>
+      </ul>
+      <ul id="awards">
+        <li class=""><em>Awards: </em><%= @movie.Awards %></li>
+        <li class=""><em>Year: </em><%= @movie.Year %></li>
+        <li class=""><em>Metascore: </em><%= @movie.Metascore %></li>
+        <li class=""><em>IMDB Rating: </em><%= @movie.imdbRating %></li>
```
Comment (philsof) [inline app/views/movies/show.html.erb:43]: `@movie.imdb_rating`

File: app/views/movies/show.html.erb [hunk 33]
```diff
@@ -0,0 +1,164 @@
...
+    <section id="conversations">
+      <div style="text-align:center;">
+<div class="container">
+
+       <h3>Comments</h3>
+  <%if @comments%>
+    <% @comments.each do |comment|%>
+      <p><%=comment.content%></p>
+      <p><%=comment.user.name%></p>
+      <p><%=comment.created_at.to_datetime.in_time_zone("Central Time (US & Canada)").strftime("%A, %d %b %Y")%></p>
```
Comment (philsof) [inline app/views/movies/show.html.erb:100]: Move this logic into a view helper

File: app/views/movies/show.html.erb [hunk 34]
```diff
@@ -0,0 +1,164 @@
...
+        <li>By: <%=conversation.user.name%> </li>
+        <hr style="margin-left: 35%; margin-right: 35%;">
+          <%end%>
+        <%end%>
+        </ul>
+      </div>
+      <% if logged_in? %>
+        <div class="conversation-post-box">
+        <%= form_for [comment, @conversation] do |f| %>
+        <%= f.hidden_field :user_id, value: current_user.id %>
```
Comment (philsof) [inline app/views/movies/show.html.erb:122]: Don't get current_user.id from form submissions; instead call `current_user` on server side. It's not safe to get this info from form submissions, since form data submissions can be spoofed. (same goes for every other place this is being done)

File: config/database.yml [hunk 35]
```diff
@@ -14,72 +14,86 @@
...
+#   database: IMDB_production ################################
+#   username: IMDB ################################
+#   password: <%= ENV['IMDB_DATABASE_PASSWORD'] %> ################################
+development:
+  adapter: sqlite3
+  database: imdb_development.db
+
+test:
+  adapter: sqlite3
+  database: imdb_development.db
```
Comment (philsof) [inline config/database.yml:92]: Your test database is the same as your development database. Not good. Be sure to name your test database something different than your development database. (Each environment should have its own database.) Take a look at [your original config/database.yml file](https://github.com/OniDotun123/IMDB/blob/8fc702f30a32b33ac936097f6c4ec70f882c44e2/config/database.yml#L60) for an example of proper database naming for each environment.

File: config/database.yml [hunk 36]
```diff
@@ -14,72 +14,86 @@
...
+
 production:
-  <<: *default
-  database: IMDB_production
-  username: IMDB
-  password: <%= ENV['IMDB_DATABASE_PASSWORD'] %>
+  adapter: postgresql
+  database: imdb_production
+  username: username
+  password: password
```
Comment (philsof) [inline config/database.yml:98]: This password should be coming from your environment variables. Anyone who has this password can access your production database! See [your original database.yml file](https://github.com/OniDotun123/IMDB/blob/8fc702f30a32b33ac936097f6c4ec70f882c44e2/config/database.yml#L85) for an example of how to set this up so your production database is secure.

File: db/migrate/20170313144411_create_users.rb [hunk 37]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :name, null: false
+      t.string :email, null: false
```
Comment (philsof) [inline db/migrate/20170313144411_create_users.rb:5]: Definitely need some sort of uniqueness constraint on your `users` table (otherwise duplicate users will not be able to log in!)

File: db/migrate/20170313144453_create_movies.rb [hunk 38]
```diff
@@ -0,0 +1,32 @@
+class CreateMovies < ActiveRecord::Migration[5.0]
+  def change
+    create_table :movies do |t|
+      t.string :Title, null: false
```
Comment (philsof) [inline db/migrate/20170313144453_create_movies.rb:4]: Follow naming conventions: `t.string :title, null: false` (this goes for every field in this migration)

Comment (philsof) [reply db/migrate/20170313144453_create_movies.rb:4]: [Rails style guide on migrations](https://github.com/bbatsov/rails-style-guide#migrations)

File: db/migrate/20170313144513_create_likes.rb [hunk 39]
```diff
@@ -0,0 +1,10 @@
+class CreateLikes < ActiveRecord::Migration[5.0]
+  def change
+    create_table :likes do |t|
+      t.references :movie, null: false
+      t.references :user, null: false
```
Comment (philsof) [inline db/migrate/20170313144513_create_likes.rb:5]: Let's get a multi-column uniqueness constraint at the database level here. (It's not good enough to have only a model validation for this, since it is the database that is the keeper of the data).

File: db/migrate/20170313150118_create_watchlists.rb [hunk 40]
```diff
@@ -0,0 +1,10 @@
+class CreateWatchlists < ActiveRecord::Migration[5.0]
+  def change
+    create_table :watchlists do |t|
+      t.references :user, null: false
+      t.references :movie, null: false
```
Comment (philsof) [inline db/migrate/20170313150118_create_watchlists.rb:5]: Heads up: right now, each `watchlist` belongs to a specific `movie`. If this is what I think it is, you will want a many-to-many here.

File: db/migrate/20170314190306_create_favorites.rb [hunk 41]
```diff
@@ -0,0 +1,8 @@
+class CreateFavorites < ActiveRecord::Migration[5.0]
+  def change
+    create_table :favorites do |t|
+
```
Comment (philsof) [inline db/migrate/20170314190306_create_favorites.rb:4]: ? are you using this table ?

File: db/migrate/20170315152028_create_conversations.rb [hunk 42]
```diff
@@ -0,0 +1,11 @@
+class CreateConversations < ActiveRecord::Migration[5.0]
+  def change
+    create_table :conversations do |t|
+      t.string :content, null: false
+      t.references :comment, null: false
+      t.references :user, null: false
```
Comment (philsof) [inline db/migrate/20170315152028_create_conversations.rb:6]: Looks like each `conversation` belongs to a specific `comment` and a specific `user`.  The specs asked for conversations on movies. (So a conversation would have many comments, but each conversation would belong to one movie.)

File: db/schema.rb [hunk 43]
```diff
@@ -0,0 +1,93 @@
...
+ActiveRecord::Schema.define(version: 20170315152028) do
+
+  create_table "comments", force: :cascade do |t|
+    t.string   "content",    null: false
+    t.integer  "movie_id",   null: false
+    t.integer  "user_id",    null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["movie_id"], name: "index_comments_on_movie_id"
+    t.index ["user_id"], name: "index_comments_on_user_id"
```
Comment (philsof) [inline db/schema.rb:22]: 👍  FYI: `t.references` in your migrations creates these indexes under the hood, which is a best-practice for foreign keys. This is another benefit of `t.references`.

File: spec/models/comment_spec.rb [hunk 44]
```diff
@@ -0,0 +1,80 @@
...
+RSpec.describe Comment, type: :model do
+  let(:test_movie) { Movie.new }
+  let(:test_user) { User.create!(name: "Carlos", email: "Carlos@gmail.com", password: "password" ) }
+
+  describe 'validations' do
+
+    before(:each) do
+      test_movie.get_movies_by_title("Fight Club")
+      first_movie = Movie.first
+      @new_comment = first_movie.comments.new(content: "This film was a work of art", user_id: test_user.id)
```
Comment (philsof) [inline spec/models/comment_spec.rb:12]: No need for instance variables in your tests, since none of these variables are being passed around.

Comment (philsof) [reply spec/models/comment_spec.rb:12]: FYI looks like RSpec now prefers `before(:example)` instead of `before(:each)`, although both will work the same way, it appears. Docs [here](https://www.relishapp.com/rspec/rspec-core/v/3-5/docs/hooks/before-and-after-hooks).

File: spec/models/comment_spec.rb [hunk 45]
```diff
@@ -0,0 +1,80 @@
...
+    end
+
+    it 'saves a new comment if successfully created' do
+      expect{@new_comment.save}.to change{Comment.count}.by(1)
+    end
+
+    it 'saves a new comment if content is present' do
+      if @new_comment.content = !nil
```
Comment (philsof) [inline spec/models/comment_spec.rb:20]: Heads up: `!nil` will always evaluate to `true`. Thus, this line is the same as:
```ruby
if @new_comment.content = true
```

Comment (philsof) [reply spec/models/comment_spec.rb:20]: When testing, since you are in complete control of the context of each test, you should not need any conditional logic in any of your tests. Instead, have different tests for different contexts. E.g.: have one test where @new_comment.content exists, and another test where @new_comment.content does not exist. Given-->Then-->When

File: spec/models/comment_spec.rb [hunk 46]
```diff
@@ -0,0 +1,80 @@
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
+  let(:test_movie) { Movie.new }
+  let(:test_user) { User.create!(name: "Carlos", email: "Carlos@gmail.com", password: "password" ) }
+
+  describe 'validations' do
+
+    before(:each) do
+      test_movie.get_movies_by_title("Fight Club")
```
Comment (philsof) [inline spec/models/comment_spec.rb:10]: This line depends on your API code. Do you want this test to also be testing your API code?

**Review by philsof — COMMENTED** [review]
@TatianaAb @prostko @OniDotun123 @k-shetty @jordanbeth See comments for remainder of code review
